Silicon/NXP/iMXPlatformPkg/Drivers/SmbiosConfigDxe/SmbiosConfigDxe.c

```diff
@@ -0,0 +1,712 @@
+/** @file
+*
+*  Copyright (c) 2019 Microsoft Corporation. All rights reserved.
+*
+*  This program and the accompanying materials
+*  are licensed and made available under the terms and conditions of the BSD License
+*  which accompanies this distribution.  The full text of the license may be found at
+*  http://opensource.org/licenses/bsd-license.php
+*
+*  THE PROGRAM IS DISTRIBUTED UNDER THE BSD LICENSE ON AN "AS IS" BASIS,
+*  WITHOUT WARRANTIES OR REPRESENTATIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED.
+*
+**/
+
+#include "SmbiosConfigDxe.h"
+
+typedef enum JsonParserState {
+  StateFindKey,
+  StateReadKey,
+  StateKeyDone,
+  StateFindValue,
+  StateReadValue,
+  StateValueDone,  
+} JSON_PARSER_STATE;
+
+#define MAX_VARIABLE_SIZE 100
+
+typedef struct SmbiosOverrideNode {
+  CHAR8 *Key;
+  CHAR8 *Value;
+  struct SmbiosOverrideNode *Next;
+} SMBIOS_OVERRIDE_NODE;
+
+STATIC CONST CHAR16 mSmbiosOverridePresent[] = L"SmbiosOverridePresent";
+
+STATIC SMBIOS_OVERRIDE_NODE *mSmbiosOverrideListHead;
+
+/**
+
+  Finds and retrieves the next Key/Value pair from the buffer starting at the
+  given offset.
+
+  This routine scans through the provided buffer starting at the given offset.
+  If finds the first '"' character and records this location as the start of the Key.
+  Then it contineus scanning to find the next '"' instance, which denotes the KeyLength.
+
+  We repeat these steps to find the Value position and length.
+
+  At any point if a failure is detected, the algorithm bails.
+
+  If we successfully find both Key and Value positions and lengths, then allocate/copy
+  these to their respective buffers and return to the caller.
+
+**/
+EFI_STATUS
+EFIAPI
+GetNextKeyValuePair (
+  CHAR8 *Buffer,
+  UINTN *Offset,
+  UINTN BufferSize,
+  CHAR8 **Key,
+  CHAR8 **Value
+  )
+{
+  EFI_STATUS Status;
+  UINTN CurrentOffset;
+  UINTN KeyLength;
+  UINTN ValueLength;
+  JSON_PARSER_STATE State;
+  CHAR8 *KeyStart;
+  CHAR8 *ValueStart;
+  CHAR8 *TempKey;
+  CHAR8 *TempValue;
+
+  CurrentOffset = *Offset;
+  KeyLength = 0;
+  ValueLength = 0;
+  Status = EFI_NOT_FOUND;
+  State = StateFindKey;
+  KeyStart = NULL;
+  ValueStart = NULL;
+  TempKey = NULL;
+  TempValue = NULL;
+
+  while (CurrentOffset < BufferSize) {
```

Comment:
Refactor Json Parser into its own library?

Reply:
Also this simplified json is probably a bit excessive. Could achieve the same result using a simple CSV parser. Consider changing over once the parsing code is in its own library.